<a href="https://colab.research.google.com/github/asterinwl/Tensorflow/blob/master/boston_pred.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

 # 1.과거의 데이터를 준비합니다

In [3]:
import pandas as pd

In [4]:
보스턴=pd.read_csv('boston.csv')
보스턴.head()

crim    zn  indus  chas    nox  ...  tax  ptratio       b  lstat  medv
0  0.00632  18.0   2.31     0  0.538  ...  296     15.3  396.90   4.98  24.0
1  0.02731   0.0   7.07     0  0.469  ...  242     17.8  396.90   9.14  21.6
2  0.02729   0.0   7.07     0  0.469  ...  242     17.8  392.83   4.03  34.7
3  0.03237   0.0   2.18     0  0.458  ...  222     18.7  394.63   2.94  33.4
4  0.06905   0.0   2.18     0  0.458  ...  222     18.7  396.90   5.33  36.2

[5 rows x 14 columns]

In [5]:
독립=보스턴[['crim','zn','indus','chas','nox','rm','age','dis','rad','tax','ptratio','b','lstat']]
종속=보스턴[['medv']]
print(독립.shape,종속.shape)

(506, 13) (506, 1)


# 2.모델의 구조를 만듭니다

In [6]:
import tensorflow as tf

In [7]:
X=tf.keras.layers.Input(shape=[13])
Y=tf.keras.layers.Dense(1)(X)
model=tf.keras.models.Model(X,Y)
model.compile(loss='mse')

독립변수는 13개고 종속변수는 1개이므로 X는 shape가 13,Y는 Dense가 1이다.

# 3.데이터로 모델을 학습(FIT)합니다.

In [8]:
model.fit(독립,종속,epochs=10)

Epoch 1/10
16/16 [==============================] - 1s 1ms/step - loss: 8132.8564
Epoch 2/10
16/16 [==============================] - 0s 1ms/step - loss: 7309.1064
Epoch 3/10
16/16 [==============================] - 0s 975us/step - loss: 6671.7939
Epoch 4/10
16/16 [==============================] - 0s 1ms/step - loss: 6118.7119
Epoch 5/10
16/16 [==============================] - 0s 945us/step - loss: 5593.9512
Epoch 6/10
16/16 [==============================] - 0s 1ms/step - loss: 5102.9902
Epoch 7/10
16/16 [==============================] - 0s 1ms/step - loss: 4644.5283
Epoch 8/10
16/16 [==============================] - 0s 983us/step - loss: 4207.8320
Epoch 9/10
16/16 [==============================] - 0s 1ms/step - loss: 3816.3564
Epoch 10/10
16/16 [==============================] - 0s 1ms/step - loss: 3444.3718


In [11]:
model.fit(독립,종속,epochs=1000, verbose=0)

verbose=0은 결과값을 보지 않겠다는 의미이다.

# 4.모델을 이용합니다.

In [12]:
print("Predictions: ",model.predict(독립[0:5]))

Predictions:  [[29.714752]
 [25.19456 ]
 [30.614456]
 [30.023575]
 [29.338787]]


1행부터 6행까지의 medv의 값을 추측해보았다.

In [13]:
종속[0:5]

medv
0  24.0
1  21.6
2  34.7
3  33.4
4  36.2

실제 1행부터 6행까지의 medv의 값이다. 차이가 존재함을 알 수 있다.

# 5.모델의 수식 확인

In [15]:
model.get_weights()

[array([[-0.08543278],
        [ 0.08007725],
        [-0.0618079 ],
        [ 3.387237  ],
        [ 2.0354414 ],
        [ 3.6444511 ],
        [ 0.02291095],
        [-0.8729906 ],
        [ 0.14357069],
        [-0.00920379],
        [ 0.10579202],
        [ 0.01714006],
        [-0.6092196 ]], dtype=float32), array([2.6292422], dtype=float32)]

1행은 -0.08543278 이 곱해있고 2행은 0.08007725이 곱해있고... 하는 방식으로 식을 만들 수 있다.